pulsar-broker/src/main/java/org/apache/pulsar/compaction/CompactedTopicImpl.java

```diff
@@ -70,14 +70,14 @@ public CompactedTopicImpl(BookKeeper bk) {
     @Override
     public CompletableFuture<CompactedTopicContext> newCompactedLedger(Position p, long compactedLedgerId) {
         synchronized (this) {
-            compactionHorizon = (PositionImpl) p;
-
             CompletableFuture<CompactedTopicContext> previousContext = compactedTopicContext;
             compactedTopicContext = openCompactedLedger(bk, compactedLedgerId);
 
+            compactionHorizon = (PositionImpl) p;
```

Comment:
We can not guarantee that the reading of `compactionHorizon` and `compactedTopicContext` safety even if this patch is merged. We should add some code-comment on the method `getCompactedTopicContextFuture` and `getCompactedTopicContext` like this:

```
/**
 * If you want to guarantee that `compactionHorizon` and `compactedTopicContext` match, you need to call this method in "synchronized(compartor){ ... }" lock block
 */
```

Reply:
I think we don't need to give it a strong guarantee. :)

Reply:
I have added some comments to explain this.